zeroex/orderwatch/order_watcher.go

```diff
@@ -539,6 +587,48 @@ func (w *Watcher) Add(orderInfo *ordervalidator.AcceptedOrderInfo) error {
 	return nil
 }
 
+func (w *Watcher) trimOrdersAndFireEvents() error {
+	targetMaxOrders := int(maxOrdersTrimRatio * float64(w.maxOrders))
+	newMaxExpirationTime, removedOrders, err := w.meshDB.TrimOrdersByExpirationTime(targetMaxOrders)
+	if err != nil {
+		return err
+	}
+	if len(removedOrders) > 0 {
+		logger.WithFields(logger.Fields{
+			"numOrdersRemoved": len(removedOrders),
+			"targetMaxOrders":  targetMaxOrders,
+		}).Debug("removing orders to make space")
+	}
+	for _, removedOrder := range removedOrders {
+		// Fire a "REMOVED" event for each order that was removed.
+		orderEvent := &zeroex.OrderEvent{
+			OrderHash:                removedOrder.Hash,
+			SignedOrder:              removedOrder.SignedOrder,
+			FillableTakerAssetAmount: removedOrder.FillableTakerAssetAmount,
+			EndState:                 zeroex.ESOrderRemoved,
+		}
+		w.orderFeed.Send([]*zeroex.OrderEvent{orderEvent})
+		// TODO(albrow): remove order from in-memory state.
```

Comment:
Removing in-memory state involves:

```
expirationTimestamp := time.Unix(order.SignedOrder.ExpirationTimeSeconds.Int64(), 0)
w.expirationWatcher.Remove(expirationTimestamp, order.Hash.Hex())

err = w.removeAssetDataAddressFromEventDecoder(order.SignedOrder.MakerAssetData)
if err != nil {
	// This should never happen since the same error would have happened when adding
	// the assetData to the EventDecoder.
	logger.WithFields(logger.Fields{
		"error":       err.Error(),
		"signedOrder": order.SignedOrder,
	}).Error("Unexpected error when trying to remove an assetData from decoder")
	return err
}
```